app/assets/javascripts/3dbio_viewer/src/webapp/components/molecular-structure/usePluginRef.ts

```diff
@@ -177,11 +276,87 @@ export function usePluginRef(options: Options) {
+                            i18n.t("Loading EMDB...")
+                        );
+                        setEmdbOpacity({ plugin, id: emdbId, value: 0.5 });
+                    } else
+                        plugin.canvas.showToast({
+                            title: i18n.t("Error"),
+                            message: getErrorByStatus(emdbId, res.status),
+                            key: "init",
+                        });
+                });
+            }
+
+            async function loadWithBothPdbEmdb(
+                pdbId: string,
+                emdbId: string,
+                element: HTMLDivElement
+            ) {
+                await checkModelUrl(pdbId, "pdb")
+                    .then(res => {
+                        if (res.loaded) {
+                            return plugin
+                                .render(element, initParams)
+                                .then(() =>
+                                    newSelection.ligandId === undefined
+                                        ? loadEmdbModel(emdbId, plugin)
+                                        : Promise.resolve()
+                                )
+                                .then(() => setVisibilityAfterLoad(initParams, plugin));
+                        } else loadVoidMolstar(getErrorByStatus(pdbId, res.status));
+                    })
+                    .catch(err => loadVoidMolstar(err));
+            }
+
+            if (
+                chainChanged &&
+                pluginAlreadyRendered &&
+                newSelection.chainId &&
+                !newSelection.ligandId
+            ) {
+                plugin.visual.updateChain(newSelection.chainId);
+            } else if (pluginAlreadyRendered) {
                 //When ligand has changed
                 molstarState.current = MolstarStateActions.fromInitParams(initParams, newSelection);
-                await updateLoader("updateVisualPlugin", plugin.visual.update(initParams));
+                await updateLoader("updateVisualPlugin", plugin.visual.update(initParams))
+                    .then(() => {
+                        const items = getCurrentItems(plugin);
+                        if (items.some(item => item.type === "emdb" && item.id === emdbId))
+                            return Promise.resolve();
+
+                        return emdbId && newSelection.ligandId === undefined
+                            ? loadEmdbModel(emdbId, plugin)
+                            : Promise.resolve();
+                    })
+                    .then(() => setVisibilityAfterLoad(initParams, plugin));
+                if (newSelection.ligandId && newSelection.chainId) {
+                    console.debug("Updating ligand in molstar sequence", newSelection);
+                    plugin.visual.updateLigand({
+                        ligandId: newSelection.ligandId,
+                        chainId: newSelection.chainId,
+                    });
+                }
```

Comment:
This method is already tooo long. At least, let's abstract new code to separate functions with declarative names.